File: Data Structures & Algorithms/reorder-linked-list/submission-3.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    void reorderList(ListNode* head) {
        if (head == nullptr|| head->next == nullptr) return;

        auto slow=head;
        auto fast=head->next;
        while (fast !=nullptr){
            fast=fast->next;
            slow = slow->next;
            if (fast!=nullptr) fast=fast->next;
        }

        ListNode* prev = nullptr;
        auto second_head = slow->next;
        slow->next = nullptr;

        while (second_head!=nullptr){
            auto nxt = second_head->next;
            second_head->next = prev;
            prev = second_head;
            second_head = nxt;
        }

        auto traversal_node = head;
        while (prev!=nullptr){
            auto left_next = traversal_node->next;
            traversal_node->next = prev;
            traversal_node=traversal_node->next;
            if (left_next==nullptr){
                return;
            }
            prev = prev->next;
            traversal_node->next = left_next;
            traversal_node = traversal_node->next;
            }

        
        return;

    }
};

File: Data Structures & Algorithms/merge-two-sorted-linked-lists/submission-3.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        # if list1 == None:
        #     return list2
        # if list2 == None:
        #     return list1

        # if list1.val > list2.val:
        #     head = ListNode(list2.val)
        #     list2= list2.next
        # else:
        #     head = ListNode(list1.val)
        #     list1 = list1.next
        temp_head = head = ListNode()

        while list2 != None and list1!= None:
            if list1.val > list2.val:
                head.next = list2
                list2 = list2.next
            else:
                head.next = list1
                list1 = list1.next
            head = head.next
        head.next = list1 or list2
        return temp_head.next